subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestEventSerializer.java

```diff
@@ -183,12 +182,19 @@ public DefaultTestFailure read(Decoder decoder) throws Exception {
             for (int i = 0; i < numOfCauses; i++) {
                 causes.add(read(decoder));
             }
+
+            String rawFailureName = decoder.readString();
+            Throwable rawFailure;
+            try {
+               rawFailure = throwableSerializer.read(decoder);
+            } catch(Exception e) {
+                rawFailure = new IllegalStateException("An exception of type " + rawFailureName + " was thrown by the test, but there was a failure while transmitting the exception back to the build process", e);
```

Comment:
❕ I (weakly) don't think this is an appropriate place for an `ISE` - exactly what _state_ is illegal here/or what method has been invoked at the wrong time?  

I think a custom exception class would make it clear this problem has been anticipated by Gradle.  Or, alternately, if you think that is too heavyweight, maybe use `NotSerializableException` (also not quite a perfect fit, but perhaps closer than `ISE`)?

Reply:
Well the "illegal state" is that we don't have the exception that we want to put in the test failure.  It's admittedly a weak association.  I changed it to a custom exception class.